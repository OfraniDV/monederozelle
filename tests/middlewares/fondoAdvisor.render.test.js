'use strict';

const { renderAdvice, createFmtCupUsdPair } = require('../../middlewares/fondoAdvisor');

function deepMerge(target = {}, source = {}) {
  const output = Array.isArray(target) ? [...target] : { ...target };
  Object.keys(source).forEach((key) => {
    const value = source[key];
    if (Array.isArray(value)) {
      output[key] = value.slice();
    } else if (value && typeof value === 'object') {
      output[key] = deepMerge(output[key] || {}, value);
    } else {
      output[key] = value;
    }
  });
  return output;
}

function buildBaseResult(overrides = {}) {
  const base = {
    activosCup: 250000,
    deudasCup: -80000,
    netoCup: 170000,
    cushionTarget: 120000,
    needCup: 30000,
    disponibles: 90000,
    plan: {
      sellTarget: {
        usd: 200,
        cupIn: 90000,
      },
      sellNow: {
        usd: 150,
        cupIn: 67500,
        minWarning: false,
      },
      remainingCup: 22500,
      remainingUsd: 70,
      sellNet: 450,
      usedSellSource: 'db',
    },
    projection: {
      negativosPost: -5000,
      colchonPost: 110000,
    },
    liquidityByBank: {
      BANDEC: 50000,
      BPA: 30000,
      METRO: 20000,
    },
    config: {
      minKeepUsd: 50,
      sellRate: 450,
      minSellUsd: 40,
      liquidityBanks: ['BANDEC', 'BPA', 'METRO'],
      sellFeePct: 0.02,
      fxMarginPct: 0,
      sellRoundToUsd: 1,
      buyRateCup: 325,
      buyRateSource: 'db',
      sellRateSource: 'db',
      limitMonthlyDefaultCup: 120000,
      limitMonthlyBpaCup: 150000,
      allocationBankOrder: ['BANDEC', 'BPA', 'METRO'],
    },
    deudaAbs: 80000,
    urgency: '🟠 PRIORITARIO',
    monthlyLimits: {
      cards: [
        {
          bank: 'BANDEC',
          mask: '#1111',
          numero: '1111111111111111',
          usedOut: 30000,
          balancePos: 10000,
          remaining: 90000,
          depositCap: 80000,
          status: 'OK',
          extendable: false,
          isBolsa: false,
        },
        {
          bank: 'BPA',
          mask: '#2222',
          numero: '2222222222222222',
          usedOut: 40000,
          balancePos: 5000,
          remaining: 110000,
          depositCap: 105000,
          status: 'EXTENDABLE',
          extendable: true,
          isBolsa: false,
        },
        {
          bank: 'METRO',
          mask: '#3333',
          numero: '3333333333333333',
          usedOut: 20000,
          balancePos: 0,
          remaining: 100000,
          depositCap: 100000,
          status: 'OK',
          extendable: false,
          isBolsa: false,
        },
      ],
      totals: {
        blocked: 0,
        extendable: 1,
        totalRemaining: 300000,
      },
    },
    distributionNow: {
      assignments: [
        {
          bank: 'BANDEC',
          numero: '1111111111111111',
          mask: '#1111',
          assignCup: 40000,
          remainingAntes: 60000,
          remainingDespues: 20000,
          status: 'OK',
          isBolsa: false,
        },
        {
          bank: 'BPA',
          numero: '2222222222222222',
          mask: '#2222',
          assignCup: 20000,
          remainingAntes: 50000,
          remainingDespues: 30000,
          status: 'EXTENDABLE',
          isBolsa: false,
        },
      ],
      leftover: 5000,
      totalAssigned: 60000,
    },
    distributionTarget: {
      assignments: [
        {
          bank: 'BANDEC',
          numero: '1111111111111111',
          mask: '#1111',
          assignCup: 50000,
          remainingAntes: 60000,
          remainingDespues: 10000,
          status: 'OK',
          isBolsa: false,
        },
        {
          bank: 'BPA',
          numero: '2222222222222222',
          mask: '#2222',
          assignCup: 30000,
          remainingAntes: 60000,
          remainingDespues: 30000,
          status: 'EXTENDABLE',
          isBolsa: false,
        },
        {
          bank: 'MITRANSFER',
          numero: '4444444444444444',
          mask: '#4444',
          assignCup: 10000,
          remainingAntes: 20000,
          remainingDespues: 10000,
          status: 'OK',
          isBolsa: true,
        },
      ],
      leftover: 10000,
      totalAssigned: 90000,
    },
    buyRateCup: 325,
    buyRateSource: 'db',
    sellRateSource: 'db',
    usdInventory: 260,
  };

  return deepMerge(base, overrides);
}

describe('createFmtCupUsdPair', () => {
  test('formatea con tasa de compra', () => {
    const fmt = createFmtCupUsdPair({ hasBuyRate: true, resolvedBuyRate: 325 });
    expect(fmt(650)).toBe('650 CUP (≈ 2.00 USD)');
    expect(fmt(-325)).toBe('-325 CUP (≈ -1.00 USD)');
  });

  test('omite USD cuando no hay tasa', () => {
    const fmt = createFmtCupUsdPair({ hasBuyRate: false, resolvedBuyRate: 0 });
    expect(fmt(5000)).toBe('5,000 CUP');
    expect(fmt(Number.NaN)).toBe('0 CUP');
  });
});

describe('renderAdvice formatting', () => {
  let logSpy;

  beforeEach(() => {
    logSpy = jest.spyOn(console, 'log').mockImplementation(() => {});
  });

  afterEach(() => {
    logSpy.mockRestore();
  });

  test('incluye pares CUP/USD y nuevo orden en Estado y Objetivo', () => {
    const blocks = renderAdvice(buildBaseResult());
    const estado = blocks.find((block) => block.startsWith('📊 <b>Estado actual</b>'));
    expect(estado).toContain('Activos: 250,000 CUP (≈ 769.23 USD)');
    expect(estado).toContain('Deudas: -80,000 CUP (≈ -246.15 USD)');
    const colchIndex = estado.indexOf('Colchón actual');
    const netoIndex = estado.indexOf('Neto');
    expect(colchIndex).toBeGreaterThan(0);
    expect(netoIndex).toBeGreaterThan(colchIndex);

    const objetivo = blocks.find((block) => block.startsWith('🎯 <b>Objetivo</b>'));
    expect(objetivo).toContain('Colchón objetivo: 120,000 CUP (≈ 369.23 USD)');
    expect(objetivo).toContain('Necesidad adicional: 30,000 CUP (≈ 92.31 USD)');
  });

  test('agrega columna ≈USD(LIBRE) y totales en USD en la tabla de límites', () => {
    const blocks = renderAdvice(buildBaseResult());
    const tableIndex = blocks.findIndex((block) => block.includes('🚦 <b>Límite mensual por tarjeta</b>'));
    const tablePre = blocks[tableIndex + 2];
    expect(tablePre).toContain('≈USD(LIBRE)');
    expect(tablePre).toContain('≈USD totales: SAL');
  });

  test('muestra nota y omite pares USD cuando no hay tasa', () => {
    const blocks = renderAdvice(
      buildBaseResult({
        buyRateCup: null,
        buyRateSource: 'none',
        config: { buyRateCup: null, buyRateSource: 'none' },
      })
    );
    const joined = blocks.join('\n\n');
    expect(joined).not.toContain('(≈ ');
    expect(joined).toContain('ℹ️ No se mostró equivalente en USD porque no hay tasa de compra configurada.');
  });
});

describe('renderAdvice snapshots', () => {
  let logSpy;

  beforeEach(() => {
    logSpy = jest.spyOn(console, 'log').mockImplementation(() => {});
  });

  afterEach(() => {
    logSpy.mockRestore();
  });

  test('caso base con tasa 325', () => {
    const blocks = renderAdvice(buildBaseResult());
    expect(blocks.join('\n\n')).toMatchSnapshot();
  });

  test('caso sin tasa de compra', () => {
    const blocks = renderAdvice(
      buildBaseResult({
        buyRateCup: null,
        buyRateSource: 'none',
        config: { buyRateCup: null, buyRateSource: 'none' },
      })
    );
    expect(blocks.join('\n\n')).toMatchSnapshot();
  });

  test('caso con escenarios y leftover destacados', () => {
    const blocks = renderAdvice(
      buildBaseResult({
        plan: {
          sellTarget: { usd: 250, cupIn: 112500 },
          sellNow: { usd: 180, cupIn: 81000, minWarning: true },
          remainingCup: 31500,
          remainingUsd: 97,
          sellNet: 450,
        },
        distributionNow: {
          assignments: [
            {
              bank: 'BANDEC',
              numero: '1111111111111111',
              mask: '#1111',
              assignCup: 50000,
              remainingAntes: 60000,
              remainingDespues: 10000,
              status: 'OK',
              isBolsa: false,
            },
          ],
          leftover: 30000,
          totalAssigned: 50000,
        },
      })
    );
    expect(blocks.join('\n\n')).toMatchSnapshot();
  });
});
