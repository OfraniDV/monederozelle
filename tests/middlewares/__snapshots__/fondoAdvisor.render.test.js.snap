// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`renderAdvice snapshots caso base con tasa 325 1`] = `
"🧮 <b>Asesor de Fondo</b>

🟠 PRIORITARIO

━━━━━━━━━━━━━━━━━━

📊 <b>Estado actual</b>
• Activos: 250,000 CUP (≈ 769.23 USD)
• Deudas: -80,000 CUP (≈ -246.15 USD)
• Colchón actual: 90,000 CUP (≈ 276.92 USD)
• Neto: 170,000 CUP (≈ 523.08 USD)

ℹ️ <b>Indicadores</b>
• Deuda/Activos: 32.0%
• Avance colchón: 75% ▓▓▓▓▓▓▓▓░░
• Meses cubiertos (colchón / gasto mensual): 1.00

🎯 <b>Objetivo</b>
• Colchón objetivo: 120,000 CUP (≈ 369.23 USD)
• Necesidad adicional: 30,000 CUP (≈ 92.31 USD)

💵 <b>Inventario USD/Zelle</b>
• Total: 260 USD
• Disponible ahora: 210 USD


💸 <b>Venta requerida (Zelle)</b>
👉 Objetivo: vender 200 USD a 450 ⇒ +90,000 CUP (≈ 200.00 USD)
👉 Vende ahora: 150 USD ⇒ +67,500 CUP (≈ 150.00 USD)
• Faltante tras venta: 22,500 CUP (≈ 70.00 USD)



💠 <b>Compra sugerida (USD)</b>
• Exceso sobre colchón/deudas: 50,000 CUP (≈ 153.85 USD)
👇 Objetivo: comprar 153 USD a 325 ⇒ −49,725 CUP (≈ 153.00 USD)
👇 Compra ahora: 153 USD ⇒ −49,725 CUP (≈ 153.00 USD)


────────────────────────────────

🚦 <b>Límite mensual por tarjeta</b>

ℹ️ Límite mensual: Estándar 120,000 CUP (≈ 369.23 USD) • BPA 150,000 CUP (≈ 461.54 USD) (ampliable)

<pre>Banco  Tarjeta      SAL    SALDO    LIBRE      CAP Estado ≈USD(LIBRE)
─────────────────────────────────────────────────────────────────────
BANDEC #1111    30,000   10,000   90,000   80,000 🟢      277
BPA    #2222    40,000    5,000  110,000  105,000 🟡 ampliable      338
METRO  #3333    20,000        0  100,000  100,000 🟢      308
─────────────────────────────────────────────────────────────────────
TOTAL  —        90,000   15,000  300,000  285,000      923
≈USD totales: SAL 276.92 • SALDO 46.15 • LIBRE 923.08 • CAP 876.92</pre>

📍 <b>Sugerencia de destino</b>

<pre>Venta AHORA: 67,500 CUP (≈ 207.69 USD)
  Banco  Tarjeta   ↦ CUP cap antes→desp    Estado
  ───────────────────────────────────────────────
  BANDEC #1111   40,000   60,000→  20,000 🟢
  BPA    #2222   20,000   50,000→  30,000 🟡
  ⚠️ CUP sin destino: 5,000 CUP (≈ 15.38 USD)
Objetivo total: 90,000 CUP (≈ 276.92 USD)
  Banco  Tarjeta   ↦ CUP cap antes→desp    Estado
  ───────────────────────────────────────────────
  BANDEC #1111   50,000   60,000→  10,000 🟢
  BPA    #2222   30,000   60,000→  30,000 🟡
  MITRANSFER #4444   10,000   20,000→  10,000 🟢 🧳 fallback
  ⚠️ CUP sin destino: 10,000 CUP (≈ 30.77 USD)

  • Nota: BOLSA usada como 2º plano. Entre bolsas: -1 CUP; Bolsa→Banco: neto ≈ asignado × 0.95</pre>

────────────────────────────────

🧾 <b>Proyección post-venta</b>
• Negativos: -5,000 CUP (≈ -15.38 USD)
• Colchón proyectado: 110,000 CUP (≈ 338.46 USD) &lt; 120,000 CUP (≈ 369.23 USD)

🏦 <b>Liquidez rápida disponible</b>
• BANDEC: 50,000 CUP (≈ 153.85 USD)
• BPA: 30,000 CUP (≈ 92.31 USD)
• METRO: 20,000 CUP (≈ 61.54 USD)

📐 <b>Explicación</b>
• Fórmula: necesidad = |deudas| + colchón − activos = 80,000 + 120,000 − 250,000 = 30,000 CUP
• Objetivo USD = ceil(necesidad / sellNet) (+ márgenes y redondeo)
• sellNet = floor(SELL × (1 − fee)) ⇒ SELL: 450 (fuente DB)  fee: 2.00%
• Conversión: USD ≈ CUP / 325.00 (fuente DB)"
`;

exports[`renderAdvice snapshots caso con escenarios y leftover destacados 1`] = `
"🧮 <b>Asesor de Fondo</b>

🟠 PRIORITARIO

━━━━━━━━━━━━━━━━━━

📊 <b>Estado actual</b>
• Activos: 250,000 CUP (≈ 769.23 USD)
• Deudas: -80,000 CUP (≈ -246.15 USD)
• Colchón actual: 90,000 CUP (≈ 276.92 USD)
• Neto: 170,000 CUP (≈ 523.08 USD)

ℹ️ <b>Indicadores</b>
• Deuda/Activos: 32.0%
• Avance colchón: 75% ▓▓▓▓▓▓▓▓░░
• Meses cubiertos (colchón / gasto mensual): 1.00

🎯 <b>Objetivo</b>
• Colchón objetivo: 120,000 CUP (≈ 369.23 USD)
• Necesidad adicional: 30,000 CUP (≈ 92.31 USD)

💵 <b>Inventario USD/Zelle</b>
• Total: 260 USD
• Disponible ahora: 210 USD


💸 <b>Venta requerida (Zelle)</b>
👉 Objetivo: vender 250 USD a 450 ⇒ +112,500 CUP (≈ 250.00 USD)
👉 Vende ahora: 180 USD ⇒ +81,000 CUP (≈ 180.00 USD)
• Faltante tras venta: 31,500 CUP (≈ 97.00 USD)



💠 <b>Compra sugerida (USD)</b>
• Exceso sobre colchón/deudas: 50,000 CUP (≈ 153.85 USD)
👇 Objetivo: comprar 153 USD a 325 ⇒ −49,725 CUP (≈ 153.00 USD)
👇 Compra ahora: 153 USD ⇒ −49,725 CUP (≈ 153.00 USD)


────────────────────────────────

🚦 <b>Límite mensual por tarjeta</b>

ℹ️ Límite mensual: Estándar 120,000 CUP (≈ 369.23 USD) • BPA 150,000 CUP (≈ 461.54 USD) (ampliable)

<pre>Banco  Tarjeta      SAL    SALDO    LIBRE      CAP Estado ≈USD(LIBRE)
─────────────────────────────────────────────────────────────────────
BANDEC #1111    30,000   10,000   90,000   80,000 🟢      277
BPA    #2222    40,000    5,000  110,000  105,000 🟡 ampliable      338
METRO  #3333    20,000        0  100,000  100,000 🟢      308
─────────────────────────────────────────────────────────────────────
TOTAL  —        90,000   15,000  300,000  285,000      923
≈USD totales: SAL 276.92 • SALDO 46.15 • LIBRE 923.08 • CAP 876.92</pre>

📍 <b>Sugerencia de destino</b>

<pre>Venta AHORA: 81,000 CUP (≈ 249.23 USD)
  Banco  Tarjeta   ↦ CUP cap antes→desp    Estado
  ───────────────────────────────────────────────
  BANDEC #1111   50,000   60,000→  10,000 🟢
  ⚠️ CUP sin destino: 30,000 CUP (≈ 92.31 USD)
Objetivo total: 112,500 CUP (≈ 346.15 USD)
  Banco  Tarjeta   ↦ CUP cap antes→desp    Estado
  ───────────────────────────────────────────────
  BANDEC #1111   50,000   60,000→  10,000 🟢
  BPA    #2222   30,000   60,000→  30,000 🟡
  MITRANSFER #4444   10,000   20,000→  10,000 🟢 🧳 fallback
  ⚠️ CUP sin destino: 10,000 CUP (≈ 30.77 USD)

  • Nota: BOLSA usada como 2º plano. Entre bolsas: -1 CUP; Bolsa→Banco: neto ≈ asignado × 0.95</pre>

────────────────────────────────

🧾 <b>Proyección post-venta</b>
• Negativos: -5,000 CUP (≈ -15.38 USD)
• Colchón proyectado: 110,000 CUP (≈ 338.46 USD) &lt; 120,000 CUP (≈ 369.23 USD)

🏦 <b>Liquidez rápida disponible</b>
• BANDEC: 50,000 CUP (≈ 153.85 USD)
• BPA: 30,000 CUP (≈ 92.31 USD)
• METRO: 20,000 CUP (≈ 61.54 USD)

📐 <b>Explicación</b>
• Fórmula: necesidad = |deudas| + colchón − activos = 80,000 + 120,000 − 250,000 = 30,000 CUP
• Objetivo USD = ceil(necesidad / sellNet) (+ márgenes y redondeo)
• sellNet = floor(SELL × (1 − fee)) ⇒ SELL: 450 (fuente DB)  fee: 2.00%
• Conversión: USD ≈ CUP / 325.00 (fuente DB)"
`;

exports[`renderAdvice snapshots caso sin tasa de compra 1`] = `
"🧮 <b>Asesor de Fondo</b>

🟠 PRIORITARIO

━━━━━━━━━━━━━━━━━━

📊 <b>Estado actual</b>
• Activos: 250,000 CUP
• Deudas: -80,000 CUP
• Colchón actual: 90,000 CUP
• Neto: 170,000 CUP

ℹ️ <b>Indicadores</b>
• Deuda/Activos: 32.0%
• Avance colchón: 75% ▓▓▓▓▓▓▓▓░░
• Meses cubiertos (colchón / gasto mensual): 1.00

🎯 <b>Objetivo</b>
• Colchón objetivo: 120,000 CUP
• Necesidad adicional: 30,000 CUP

💵 <b>Inventario USD/Zelle</b>
• Total: 260 USD
• Disponible ahora: 210 USD


💸 <b>Venta requerida (Zelle)</b>
👉 Objetivo: vender 200 USD a 450 ⇒ +90,000 CUP
👉 Vende ahora: 150 USD ⇒ +67,500 CUP
• Faltante tras venta: 22,500 CUP


────────────────────────────────

🚦 <b>Límite mensual por tarjeta</b>

ℹ️ Límite mensual: Estándar 120,000 CUP • BPA 150,000 CUP (ampliable)

<pre>Banco  Tarjeta      SAL    SALDO    LIBRE      CAP Estado
─────────────────────────────────────────────────────────
BANDEC #1111    30,000   10,000   90,000   80,000 🟢
BPA    #2222    40,000    5,000  110,000  105,000 🟡 ampliable
METRO  #3333    20,000        0  100,000  100,000 🟢
─────────────────────────────────────────────────────────
TOTAL  —        90,000   15,000  300,000  285,000</pre>

📍 <b>Sugerencia de destino</b>

<pre>Venta AHORA: 67,500 CUP
  Banco  Tarjeta   ↦ CUP cap antes→desp    Estado
  ───────────────────────────────────────────────
  BANDEC #1111   40,000   60,000→  20,000 🟢
  BPA    #2222   20,000   50,000→  30,000 🟡
  ⚠️ CUP sin destino: 5,000 CUP
Objetivo total: 90,000 CUP
  Banco  Tarjeta   ↦ CUP cap antes→desp    Estado
  ───────────────────────────────────────────────
  BANDEC #1111   50,000   60,000→  10,000 🟢
  BPA    #2222   30,000   60,000→  30,000 🟡
  MITRANSFER #4444   10,000   20,000→  10,000 🟢 🧳 fallback
  ⚠️ CUP sin destino: 10,000 CUP

  • Nota: BOLSA usada como 2º plano. Entre bolsas: -1 CUP; Bolsa→Banco: neto ≈ asignado × 0.95</pre>

────────────────────────────────

🧾 <b>Proyección post-venta</b>
• Negativos: -5,000 CUP
• Colchón proyectado: 110,000 CUP &lt; 120,000 CUP

🏦 <b>Liquidez rápida disponible</b>
• BANDEC: 50,000 CUP
• BPA: 30,000 CUP
• METRO: 20,000 CUP

📐 <b>Explicación</b>
• Fórmula: necesidad = |deudas| + colchón − activos = 80,000 + 120,000 − 250,000 = 30,000 CUP
• Objetivo USD = ceil(necesidad / sellNet) (+ márgenes y redondeo)
• sellNet = floor(SELL × (1 − fee)) ⇒ SELL: 450 (fuente DB)  fee: 2.00%

ℹ️ No se mostró equivalente en USD porque no hay tasa de compra configurada."
`;
